compiler/driver.ml

```diff
@@ -16,21 +16,206 @@
    the License. *)
 
 open Catala_utils
+open Shared_ast
 
 (** Associates a file extension with its corresponding {!type: Cli.backend_lang}
     string representation. *)
 let extensions = [".catala_fr", "fr"; ".catala_en", "en"; ".catala_pl", "pl"]
 
-type backend = [ Cli.backend_option | `Plugin of Plugin.handler ]
+let modname_of_file f =
+  (* Fixme: make this more robust *)
+  String.capitalize_ascii Filename.(basename (remove_extension f))
+
+let get_lang options filename =
+  Option.bind
+    (List.assoc_opt (Filename.extension filename) extensions)
+    (fun l -> List.assoc_opt l Cli.languages)
+  |> function
+  | Some lang -> lang
+  | None -> (
+    match options.Cli.language with
+    | Some lang -> lang
+    | None ->
+      Message.raise_error
+        "Could not infer language variant from the extension of \
+         @{<yellow>%s@}, and @{<bold>--language@} was not specified"
+        filename)
+
+let load_module_interfaces prg options link_modules =
+  List.fold_left
+    (fun prg f ->
+      let lang = get_lang options f in
+      let modname = modname_of_file f in
+      Surface.Parser_driver.add_interface (FileName f) lang [modname] prg)
+    prg link_modules
+
+module Passes = struct
+  (* Each pass takes only its cli options, then calls upon its dependent passes
+     (forwarding their options as needed) *)
+
+  let surface options : Surface.Ast.program * Cli.backend_lang =
+    Message.emit_debug "Reading files...";
+    let language =
+      get_lang options
+        (match options.input_file with FileName s -> s | Contents _ -> "")
+    in
+    let prg =
+      Surface.Parser_driver.parse_top_level_file options.input_file language
+    in
+    Surface.Fill_positions.fill_pos_with_legislative_info prg, language
+
+  let desugared options ~link_modules :
+      Desugared.Ast.program * Desugared.Name_resolution.context =
+    let prg, _ = surface options in
+    let prg = load_module_interfaces prg options link_modules in
+    Message.emit_debug "Name resolution...";
+    let ctx = Desugared.Name_resolution.form_context prg in
+    (* let scope_uid = get_scope_uid options backend ctx in
+     * (\* This uid is a Desugared identifier *\)
+     * let variable_uid = get_variable_uid options backend ctx scope_uid in *)
+    Message.emit_debug "Desugaring...";
+    let prg = Desugared.From_surface.translate_program ctx prg in
+    Message.emit_debug "Disambiguating...";
+    let prg = Desugared.Disambiguate.program prg in
+    Message.emit_debug "Linting...";
+    Desugared.Linting.lint_program prg;
+    prg, ctx
+  (* Note: we forward the name resolution context throughout in order to locate
+     uids from strings. Maybe a reduced form should be included directly in
+     [prg] for that purpose *)
+
+  let scopelang options ~link_modules :
```

Comment:
This approach resulted in a much simpler interface (see the `Commands` module: one function call to get to the expected AST instead of repeating the same sequence and passing the proper options and results along) ; also function calls allow for a more flexible call graph and this feels easier to read overall ; the situation could be different if we had passes with similar types that could be optional or reordered, but such cases are limited in Catala.